docs/proposals/scheduling/20240131-pod-level-numa-policy.md

```diff
@@ -0,0 +1,281 @@
+
+[//]: # (- `besteffort`: a numa with a SingleNUMANode pod will not be scheduled another pod with multi-numa if there is 
+another idle numa.)
+
+[//]: # (- `exclusive`: a numa with a SingleNUMANode pod can not be scheduled another pod with multi-numa.)
+
+[//]: # (`SingleNUMANodeExclusive` should not be set by users, because some users may set all his/her pods as exclusive 
+so that all pods with restricted policy cannot be scheduled.)
+
+#### Examples
+``` yaml
+metadata:
+  annotations:|-
+      {
+        "numaTopologyPolicy": "SingleNUMANode",
+      }
+spec:
+  containers:
+    resource:
+      requests:
+        cpu: 1
+      limits:
+        cpu: 1
+```
+This pod will be allocated with an exclusive cpu.
+
+``` yaml
+metadata:
+  annotations:|-
+      {
+        "numaTopologyPolicy": "SingleNUMANode",
+      }
+spec:
+  containers:
+    resource:
+      requests:
+        cpu: 1
+      limits:
+        cpu: 2
+```
+This pod will be allocated in shared pool.
+
+``` yaml
+metadata:
+  annotations:|-
+      {
+        "numaTopologyPolicy": "SingleNUMANode",
+      }
+spec:
+  containers:
+    resource:
+      requests:
+        cpu: 1
+        nvidia.com/gpu: 1
+      limits:
+        cpu: 1
+        nvidia.com/gpu: 1
+```
+GPU and cpu will be aligned in one NUMA node for this pod.
+
+``` yaml
+spec:
+  containers:
+    resource:
+      requests:
+        cpu: 1
+        nvidia.com/gpu: 1
+      limits:
+        cpu: 1
+        nvidia.com/gpu: 1
+```
+We will not bind align gpu and cpu for this pod in one numa, because there is no
+numa topology policy on pod.
+
+### Work with Node-Level Policy
+
+We want to make the new API compatible with the numa topology policy on node. So if the policy on pod and on node are 
+different, we will not place the pod on node. If the policy on node is `""`, means the node is able to place any workload,
+then we can schedule pod as we like. On the other hand, `""` in workload's policy means the pod do not have any requirement
+for scheduling, so it can be placed on any node. Just schedule it as we what we do before.
+
+So we have these rules:
+
+- If the policy on pod is not `""`, the scheduler will use the policy set on pod.
+Pod with a not-none policy can be scheduled on a node only if the policy
+on the node is `""` or the same as the policy on pod.
+
+- If the policy on pod is not set or is `""`, then the scheduler will use the policy set on node.
+
+|                | SingleNUMANode node | Restricted node | Besteffort node | none node | 
+|----------------|-----------------------|-----------------|-----------------|-----------|
+| SingleNUMANode | ✅                     | ❌               | ❌               | ✅         |
+| Restricted     | ❌                     | ✅               | ❌               | ✅         |
+| Besteffort     | ❌                     | ❌               | ✅               | ✅         |
+| none           | ✅                     | ✅               | ✅               | ✅         |
```

Comment:
I agree with the rules defined in this table and the principle of keeping it simple and clear. But I’m also wondering whether there are some ambiguities in it intuitively, such as: Pod+Besteffort, can it be placed on SingleNUMANode and RestrictedNode?

Reply:
By design it can not be placed on SingleNUMANode and RestrictedNode. I think bestEffort should be considered as a numa topology policy and none should be considered as without numa topology policy. A node with numa topology policy can only accept pods with same policy or without numa topology policy.